runtime/pkg/duckdbreplicator/backup.go

```diff
@@ -0,0 +1,335 @@
+	return b.bucket.Close()
+}
+
+type GCSBackupProviderOptions struct {
+	// UseHostCredentials specifies whether to use the host's default credentials.
+	UseHostCredentials         bool
+	ApplicationCredentialsJSON string
+	// Bucket is the GCS bucket to use for backups. Should be of the form `bucket-name`.
+	Bucket string
+	// BackupFormat specifies the format of the backup.
+	// TODO :: implement backup format. Fixed to DuckDB for now.
+	BackupFormat BackupFormat
+	// UnqiueIdentifier is used to store backups in a unique location.
+	// This must be set when multiple databases are writing to the same bucket.
+	UniqueIdentifier string
+}
+
+// NewGCSBackupProvider creates a new BackupProvider based on GCS.
+func NewGCSBackupProvider(ctx context.Context, opts *GCSBackupProviderOptions) (*BackupProvider, error) {
+	client, err := newClient(ctx, opts.ApplicationCredentialsJSON, opts.UseHostCredentials)
+	if err != nil {
+		return nil, err
+	}
+
+	bucket, err := gcsblob.OpenBucket(ctx, client, opts.Bucket, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open bucket %q, %w", opts.Bucket, err)
+	}
+
+	if opts.UniqueIdentifier != "" {
+		if !strings.HasSuffix(opts.UniqueIdentifier, "/") {
+			opts.UniqueIdentifier += "/"
+		}
+		bucket = blob.PrefixedBucket(bucket, opts.UniqueIdentifier)
+	}
+	return &BackupProvider{
+		bucket: bucket,
+	}, nil
+}
+
+// syncWrite syncs the write path with the backup location.
+func (d *db) syncWrite(ctx context.Context) error {
+	if !d.writeDirty || d.backup == nil {
+		// optimisation to skip sync if write was already synced
+		return nil
+	}
+	d.logger.Debug("syncing from backup")
+	// Create an errgroup for background downloads with limited concurrency.
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(8)
+
+	objects := d.backup.List(&blob.ListOptions{
+		Delimiter: "/", // only list directories with a trailing slash and IsDir set to true
+	})
+
+	tblVersions := make(map[string]string)
+	for {
+		// Stop the loop if the ctx was cancelled
+		var stop bool
+		select {
+		case <-ctx.Done():
+			stop = true
+		default:
+			// don't break
+		}
+		if stop {
+			break // can't use break inside the select
+		}
+
+		obj, err := objects.Next(ctx)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+		if !obj.IsDir {
+			continue
+		}
+
+		table := strings.TrimSuffix(obj.Key, "/")
+		d.logger.Debug("SyncWithObjectStorage: discovered table", slog.String("table", table))
+
+		// get version of the table
+		var backedUpVersion string
+		err = retry(func() error {
+			res, err := d.backup.ReadAll(ctx, filepath.Join(table, "version.txt"))
+			if err != nil {
+				return err
+			}
+			backedUpVersion = string(res)
+			return nil
+		})
+		if err != nil {
+			if gcerrors.Code(err) == gcerrors.NotFound {
+				// invalid table directory
+				d.logger.Debug("SyncWithObjectStorage: invalid table directory", slog.String("table", table))
+				_ = d.deleteBackup(ctx, table, "")
```

Comment:
There can't be a concurrent writer. It should always be called with the write mutex held. I will add a clarifying comment as well.

Reply:
Concurrent writers aren't critical, but it would be nice to handle them gracefully if possible. What if it only deletes versions that are lower than the current `version.txt`? I guess that would enable eventual garbage collection without the risk of deleting something being contributed by another writer.

Reply:
So this cleanup happens only when `version.txt` is not found so there can't be a comparison of versions. 
Ideally even concurrent ETL workers should not work for the same instance otherwise it can lead to unknown bugs (Imagine one concurrent ETL worker mutating table on which another ETL worker depends though ideally reconciler will prevent them from happening but we need to think through). 

We can skip the cleanup for now if you are worried about data losses due to such bugs.